### PR TITLE
[sailfish-browser] Allow adding new search engines

### LINKED ISSUE #254: Ad possibility to add other search engines
Please add the option to simply add other existing search engines by url.

### PR DESCRIPTION
**Pull request by siteshwar**
Fixes #254
Read opensearch descriptors from '/home/nemo/.local/share/org.sailfishos/sailfish-browser/searchEngines/'

### CODE REVIEW

File: settings/browsersettings.cpp [hunk 1]
```diff
@@ -13,7 +13,9 @@
 #include <QXmlStreamReader>
 #include "browsersettings.h"
 
-static const QString openSearchPath("/usr/lib/mozembedlite/chrome/embedlite/content/");
+static const QStringList gOpenSearchPathList = QStringList() \
+                                            << QString("/usr/lib/mozembedlite/chrome/embedlite/content/") \
+                                            << QDir::homePath() + "/.local/share/org.sailfishos/sailfish-browser/opensearch/";
```
Comment (rainemak) [inline settings/browsersettings.cpp]: Please move ```/usr/lib/mozembedlite/chrome/embedlite/content/``` to a define in common.pri (similar to DEFAULT_COMPONENTS_PATH) and replace also from browsersettings.cpp. For user specific search engine paths use 
QStandardPaths::writableLocation(QStandardPaths::DataLocation) + "/searchEngines".

File: src/declarativewebutils.cpp [hunk 2]
```diff
@@ -24,9 +24,12 @@
 #include "declarativewebutils.h"
 #include "qmozcontext.h"
 
+static QString userOpenSearchPath = QDir::homePath() + "/.local/share/org.sailfishos/sailfish-browser/opensearch/";
 static const QString gSystemComponentsTimeStamp("/var/lib/_MOZEMBED_CACHE_CLEAN_");
 static const QString gProfilePath("/.mozilla/mozembed");
-static const QString gOpenSearchPath("/usr/lib/mozembedlite/chrome/embedlite/content/");
+static const QStringList gOpenSearchPathList = QStringList() \
+                                                << QString("/usr/lib/mozembedlite/chrome/embedlite/content/") \
+                                                << userOpenSearchPath;
```
Comment (rainemak) [inline src/declarativewebutils.cpp]: Related to the first comment would be nice if we'd encapsulate search engines to a class and this way share more code between webutils and settings. Static methods for paths and for parsing xmls should be enough.

You could add search.pri under "src" directory and include it in both src.pro and settings.pro. The search.pri could then include all needed header and source files.

File: src/declarativewebutils.cpp [hunk 3]
```diff
@@ -24,9 +24,12 @@
 #include "declarativewebutils.h"
 #include "qmozcontext.h"
 
+static QString userOpenSearchPath = QDir::homePath() + "/.local/share/org.sailfishos/sailfish-browser/opensearch/";
```
Comment (rojkov) [inline src/declarativewebutils.cpp]: Global variables can be tricky. Safer to declare it as constant.

File: src/declarativewebutils.cpp [hunk 4]
```diff
@@ -411,7 +417,7 @@ void DeclarativeWebUtils::handleObserve(const QString message, const QVariant da
 
                     // load opensearch descriptions
                     loadsearch.insert(QString("msg"), QVariant(QString("loadxml")));
-                    loadsearch.insert(QString("uri"), QVariant(QString("chrome://embedlite/content/") +
+                    loadsearch.insert(QString("uri"), QVariant(QString("file://") + userOpenSearchPath +
```
Comment (rojkov) [inline src/declarativewebutils.cpp]: Are the system engines loaded (the ones from chrome://embedlite/content/)?

Comment (siteshwar) [inline src/declarativewebutils.cpp]: Yes.

File: settings/browsersettings.cpp (hunk 1 above)
Comment (siteshwar) [inline settings/browsersettings.cpp]: ```QStandardPaths::writableLocation(QStandardPaths::DataLocation) + "/searchEngines"``` will have different paths for browser application and browser settings.
For browser application it is ```/home/nemo/.local/share/jolla-settings/searchEngines```.
For browser settings it is ```/home/nemo/.local/share/org.sailfishos/sailfish-browser/searchEngines```.

Comment (rainemak) [inline settings/browsersettings.cpp]: Yeap, for sure. Would be be having a define in common.pri (one place) and using that define.

**Comment by siteshwar** [comment]
Updated PR

File: common/opensearchconfigs.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        }
+    }
+    return configs;
+}
+
+const QStringList OpenSearchConfigs::parseSearchEngineList()
+{
+    QStringList engineList;
+
+    foreach(QString openSearchPath, m_openSearchPathList) {
```
Comment (rojkov) [inline common/opensearchconfigs.cpp]: If we merge two similar functions into a single compilation unit then we should try to merge the functions into one as well. Currently the code of OpenSearchConfigs::parseSearchEngineList() and OpenSearchConfigs::parseOpenSearchConfigs() looks like code duplication. Would be better to move the common functionality to a common function.

File: src/declarativewebutils.cpp [hunk 6]
```diff
@@ -411,7 +378,7 @@ void DeclarativeWebUtils::handleObserve(const QString message, const QVariant da
 
                     // load opensearch descriptions
                     loadsearch.insert(QString("msg"), QVariant(QString("loadxml")));
-                    loadsearch.insert(QString("uri"), QVariant(QString("chrome://embedlite/content/") +
+                    loadsearch.insert(QString("uri"), QVariant(QString("file://") + OpenSearchConfigs::userOpenSearchPath() +
```
Comment (rojkov) [inline src/declarativewebutils.cpp]: This won't work. Try to delete /home/nemo/.mozilla and restart the browser: the search descriptors won't be loaded.
First `configs` will be initialized to the content of `chrome://embedlite/content/` and `registeredSearches` to an empty list. Then in this line we'll try to loaded a pre-installed search descriptor from the user folder -> fail.

Comment (siteshwar) [inline src/declarativewebutils.cpp]: Fixed through 691b6ba

File: common/opensearchconfigs.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    return openSearchConfigs;
+}
+
+const QStringList OpenSearchConfigs::getSearchEngineList()
+{
+    // Return names of search engines
+    return getInstance()->parseOpenSearchConfigs().keys();
+}
+
+QString OpenSearchConfigs::userOpenSearchPath()
```
Comment (rojkov) [inline common/opensearchconfigs.cpp]: this looks like unused method

Comment (siteshwar) [inline common/opensearchconfigs.cpp]: I left it as it might be helpful in future.

Comment (rojkov) [inline common/opensearchconfigs.cpp]: Future has proved to be unpredictable enough, so it's better to remove the code that solves the problem which doesn't exist yet :)

Comment (siteshwar) [inline common/opensearchconfigs.cpp]: Done!

File: common/opensearchconfigs.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+class OpenSearchConfigs : public QObject {
+
+public:
+    static const StringMap getAvailableOpenSearchConfigs();
+    static const QStringList getSearchEngineList();
+
+private:
+    QStringList m_openSearchPathList;
+    QString m_userOpenSearchPath;
```
Comment (rojkov) [inline common/opensearchconfigs.h]: This is redundant too.

Comment (siteshwar) [inline common/opensearchconfigs.h]: I have removed it now.

**Comment by rojkov** [comment]
LGTM

**Pull request merged**